### PR TITLE
Flow graph node caching

### PR DESCRIPTION
**Pull request by carolhmj**

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/14453/merge

Test environment: 
https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14453/merge/index.html

To test a playground add it to the URL, for example:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14453/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/14453/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/14453/merge
https://gui.babylonjs.com/?snapshot=refs/pull/14453/merge
https://nme.babylonjs.com/?snapshot=refs/pull/14453/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/14453/merge#BCU1XR#0

**Comment by bjsplat** [comment]
Visualization tests for webgl1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14453/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
Visualization tests for webgl1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14453/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
Visualization tests for webgl2 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14453/merge/testResults/webgl2/index.html

If tests were successful afterwards, this report might not be available anymore.

File: packages/dev/core/src/FlowGraph/Blocks/Data/flowGraphCachedOperationBlock.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ */
+export abstract class FlowGraphCachedOperationBlock<OutputT> extends FlowGraphBlock {
+    public readonly output: FlowGraphDataConnection<OutputT>;
+
+    constructor(outputRichType: RichType<OutputT>, config?: IFlowGraphBlockConfiguration) {
+        super(config);
+
+        this.output = this._registerDataOutput("output", outputRichType);
+    }
+    public abstract _doOperation(context: FlowGraphContext): OutputT;
```
Comment (Popov72) [inline packages/dev/core/src/FlowGraph/Blocks/Data/flowGraphCachedOperationBlock.ts:21]: I think you have to add a `@internal` tag if the method is not supposed to be visible / used by the end user?

File: packages/dev/core/src/FlowGraph/Blocks/Data/flowGraphCachedOperationBlock.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import type { IFlowGraphBlockConfiguration } from "../../flowGraphBlock";
+import { FlowGraphBlock } from "../../flowGraphBlock";
+import type { FlowGraphContext } from "../../flowGraphContext";
+import type { FlowGraphDataConnection } from "../../flowGraphDataConnection";
+import type { RichType } from "../../flowGraphRichTypes";
+
+const CACHE_NAME = "cachedOperationValue";
+const EXEC_ID_NAME = "executionId";
```
Comment (Popov72) [inline packages/dev/core/src/FlowGraph/Blocks/Data/flowGraphCachedOperationBlock.ts:8]: I would call it `cachedExecutionId` to make clear it is the executionId value related to `cachedOperationValue`.

File: packages/dev/core/src/FlowGraph/flowGraphContext.ts [hunk 3]
```diff
@@ -90,6 +90,11 @@ export class FlowGraphContext {
      * These are blocks that have currently pending tasks/listeners that need to be cleaned up.
      */
     private _pendingBlocks: FlowGraphAsyncExecutionBlock[] = [];
+    /**
+     * An monotonically increasing ID for each execution.
+     * Incremented every for every block executed.
```
Comment (Popov72) [inline packages/dev/core/src/FlowGraph/flowGraphContext.ts:95]:
```suggestion
     * Incremented for every block executed.
```

File: packages/dev/core/src/FlowGraph/flowGraphContext.ts [hunk 4]
```diff
@@ -90,6 +90,11 @@ export class FlowGraphContext {
      * These are blocks that have currently pending tasks/listeners that need to be cleaned up.
      */
     private _pendingBlocks: FlowGraphAsyncExecutionBlock[] = [];
+    /**
+     * An monotonically increasing ID for each execution.
```
Comment (Popov72) [inline packages/dev/core/src/FlowGraph/flowGraphContext.ts:94]:
```suggestion
     * A monotonically increasing ID for each execution.
```

**Review by Popov72 — APPROVED** [review]

**Review by RaananW — APPROVED** [review]

**Pull request merged**